README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# discretefit

<!-- badges: start -->

[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/discretefit)](https://cran.r-project.org/package=discretefit)
[![CRAN\_latest\_release\_date](https://www.r-pkg.org/badges/last-release/discretefit)](https://cran.r-project.org/package=discretefit)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://choosealicense.com/licenses/mit/)

<!-- badges: end -->

The package `discretefit` implements fast Monte Carlo simulations for
goodness-of-fit (GOF) tests for discrete distributions. This includes
tests based on the root-mean-square statistic, the Chi-squared
statistic, the log-likelihood-ratio (*G*<sup>2</sup>) statistic, the
Freeman-Tukey (Hellinger-distance) statistic, the Kolmogorov-Smirnov
statistic, and the Cramer-von Mises statistic.

Simulations are written in C++ (utilizing `Rcpp`) and are considerably
faster than the simulated Chi-squared GOF test in the R `stats` package.

## Installation

You can install `discretefit` from CRAN.

``` r
install.packages("discretefit")
```

You can also install the development version from GitHub:

``` r
devtools::install_github("josh-mc/discretefit")
```

## Usage

The GOF tests in `discretefit` function on a vector of counts, x, and a
vector of probabilities, p. In the below example, x represents a vector
of counts for five categories, and p represents a vector of
probabilities for each corresponding category. The GOF tests provides
p-values for the null hypothesis that x is a random sample of the
discrete distribution defined by p. 

``` r
library(discretefit)
library(bench)

x <- c(42, 0, 13, 2, 109)
p <- c(0.2, 0.05, 0.1, 0.05, 0.6)

chisq_gof(x, p)
#> 
#>  Simulated Chi-squared goodness-of-fit test
#> 
#> data:  x
#> Chi-squared = 17.082, p-value = 0.0022
rms_gof(x, p)
#> 
#>  Simulated root-mean-square goodness-of-fit test
#> 
#> data:  x
#> RMS = 1.731, p-value = 0.0389
g_gof(x, p)
#> 
#>  Simulated log-likelihood-ratio goodness-of-fit test
#> 
#> data:  x
#> G2 = 27.362, p-value = 9.999e-05
ft_gof(x, p)
#> 
#>  Simulated Freeman-Tukey goodness-of-fit test
#> 
#> data:  x
#> FT = 45.599, p-value = 9.999e-05
ks_gof(x, p)
#> 
#>  Simulated Kolmogorov-Smirnov goodness-of-fit test
#> 
#> data:  x
#> KS = 0.056627, p-value = 0.2404
cvm_gof(x, p)
#> 
#>  Simulated Cramer-von Mises goodness-of-fit test
#> 
#> data:  x
#> W2 = 0.12578, p-value = 0.1784
```

## Speed

The simulated Chi-squared GOF test in `discretefit` produces identical
p-values to the simulated Chi-squared GOF test in the `stats` package
that is part of base R.

``` r
set.seed(499)
chisq_gof(x, p, reps = 2000)$p.value
#> [1] 0.002998501
set.seed(499)
chisq.test(x, p = p, simulate.p.value = TRUE)$p.value
#> [1] 0.002998501
```

However, because Monte Carlo simulations in `discretefit` are
implemented in C++, `chisq_gof` is much faster than `chisq.test`, \~20
times faster for the following case.

``` r
speed <- bench::mark(min_iterations = 1000, check = FALSE, 
  chisq_gof = chisq_gof(x, p, reps = 2000),
  chisq.test = chisq.test(x, p = p, simulate.p.value = TRUE, B = 2000)
)
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.

ggplot2::autoplot(speed) +
  ggplot2::theme_minimal() +
  ggplot2::xlab(NULL) +
  ggplot2::ylab(NULL) 
#> Loading required namespace: tidyr
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

Additionally, the simulated GOF tests in base R is vectorized, so for
large vectors attempting a large number of simulations may not be
possible because of memory constraints. Since the functions in
`discretefit` are not vectorized, memory use is minimized.

## Root-mean-square statistic

In a surprising number of cases, a simulated GOF test based on the
root-mean-square statistic outperforms the Chi-squared test and other
tests in the Cressie-Read power divergence family. This has been
demonstrated by Perkins, Tygert, and Ward (2011). They provide the
following toy example.

Take a discrete distribution with 50 bins (or categories). The
probability for the first and second bin is 0.25. The probability for
each of the remaining 48 bins is 0.5 / 48 (\~0.0104).

Now take the observed counts of 15 for the first bin, 5 for the second
bin, and zero for each of the remaining 48 bins. It’s obvious that these
observations are very unlikely to occur for a random sample from the
above distribution. However, the Chi-squared, *G*<sup>2</sup>, and
Freeman-Tukey tests fail to reject the null hypothesis.

``` r
x <- c(15, 5, rep(0, 48))
p <- c(0.25, 0.25, rep(1/(2 * 50 -4), 48))

chisq_gof(x, p)
#> 
#>  Simulated Chi-squared goodness-of-fit test
#> 
#> data:  x
#> Chi-squared = 30, p-value = 0.9681
g_gof(x, p)
#> 
#>  Simulated log-likelihood-ratio goodness-of-fit test
#> 
#> data:  x
#> G2 = 32.958, p-value = 0.6627
ft_gof(x, p)
#> 
#>  Simulated Freeman-Tukey goodness-of-fit test
#> 
#> data:  x
#> FT = 50.718, p-value = 0.141
```

By contrast, the root-mean-square test convincingly rejects the null
hypothesis.

``` r
rms_gof(x, p)
#> 
#>  Simulated root-mean-square goodness-of-fit test
#> 
#> data:  x
#> RMS = 5.1042, p-value = 9.999e-05
```

For additional examples, see Perkins, Tygert, and Ward (2011) and Ward
and Carroll (2014).

## Alternatives

Several other packages implement GOF tests for discrete distributions.

As noted above, the `stats` package in base R implements a simulated
Chi-squared GOF test.

I’m not aware of an R package that implements a simulated
*G*<sup>2</sup> GOF test but the packages `RVAideMemoire` and
`DescTools` implement GOF tests that utilize approximations based on the
Chi-squared distribution.

The `dgof` package (Anderson and Emerson 2011) implements simulated
Kolmogorov-Smirnov GOF tests and simulated Cramer-von Mises GOF tests .
The `cvmdisc` package also implements a simulated Cramer-von Mises GOF
test.

The `KSgeneral` package (Dimitrova, Kaishev, and Tan, 2020) implements
exact Kolmogorov-Smirnov GOF tests and fast, simulated GOF tests
utilizing the algorithm introduced by Wood and Altavela (1978).

I’m not aware of another R package that implements a root-mean-square
GOF test.

## References

Arnold, Taylor B. and John W. Emerson. “Nonparametric goodness-of-fit
tests for discrete null distributions.” R Journal.
<https://doi.org/10.32614/rj-2011-016>

Dimitrova, Dimitrina S., Vladimir K. Kaishev, and Senren Tan. “Computing
the Kolmogorov-Smirnov distribution when the underlying CDF is purely
discrete, mixed, or continuous.” Journal of Statistical Software, 2020.
<https://doi.org/10.18637/jss.v095.i10>

Eddelbuettel, Dirk and Romain Francois. “Rcpp: Seamless R and C++
Integration.” Journal of Statistical Software, 2011.
<https://www.jstatsoft.org/article/view/v040i08>

Perkins, William, Mark Tygert, and Rachel Ward. “Computing the
confidence levels for a root-mean-square test of goodness-of-fit.”
Applied Mathematics and Computation, 2011.
<https://doi.org/10.1016/j.amc.2011.03.124>

Ward, Rachel and Raymond J. Carroll. “Testing Hardy–Weinberg equilibrium
with a simple root-mean-square statistic.” Biostatistics, 2014.
<https://doi.org/10.1093/biostatistics/kxt028>

Wood, Constance L., and Michele M. Altavela. “Large-Sample Results for
Kolmogorov–Smirnov Statistics for Discrete Distributions.” Biometrika,
1978. <https://doi.org/10.1093/biomet/65.1.235>


NEWS.md
# discretefit 0.1.2

* Significantly increased speed by rewriting tabulate function.

# discretefit 0.1.1

* Replaced `sqrt` with `pow` in Rcpp to pacify Solaris. 

# discretefit 0.1.0

* Changed all GOF functions to class 'htest.' This means these functions now return a list that includes the p-value as well as the test statistics. In the previous version, these functions returned a numeric value directly (the p-value).
* Added the parameter 'tolerance' to all GOF functions. This allows users directly to adjust the tolerance for numerical imprecision.
* Added the function `ft_gof()` that implements the Freeman-Tukey GOF test.
* Added the function `cvm_gof()` that implements the Cramer-von Mises GOF test.
* Updated the test requiring the vector of probabilities to sum to one to include tolerance for numerical imprecision. 
* Added a `NEWS.md` file to track changes to the package.



tests/testthat.R
library(testthat)
library(discretefit)
library(dgof)

test_check("discretefit")


tests/testthat/test-stats.R
#Defining observed and theoretical distribution
x <- c(42, 5, 13, 2, 109)
p <- c(0.2, 0.05, 0.1, 0.05, 0.6)

#for cvmdisc package
breaks <- c(0, cumsum(p))

#translation for dgof
xx <- c(rep(1, 42), rep(2, 5), rep(3, 13), rep(4, 2), rep(5, 109))
pp <- c(rep(1, 20), rep(2, 5), rep(3, 10), rep(4, 5), rep(5, 60))

test_that("chi_gof statatistic is equal to chisq.test statistic", {

  a <- as.numeric(chisq_gof(x, p, reps = 100)$statistic)

  b <- as.numeric(chisq.test(x, p = p)$statistic)

  expect_equal(a, b)

})

test_that("g_gof statistic is equal to G.test statistic (RVAideMemoire)", {

  a <- as.numeric(g_gof(x, p, reps = 100)$statistic)

  b <- 12.1450359514788
  #b <- as.numeric(RVAideMemoire::G.test(x, p)$statistic)

  expect_equal(a, b)

})


test_that("rms_gof statistic is correct", {

  a <- as.numeric(rms_gof(x, p, reps = 100)$statistic)

  #b <- sqrt(sum((x/sum(x) - p)^2) * 1/length(x))

  b <- sum((sqrt(sum(x)) * (x/sum(x) - p))^2)

  expect_equal(a, b)

})

test_that("ft_gof statistic is correct", {

  a <- as.numeric(ft_gof(x, p, reps = 100)$statistic)

  b <- 4 * sum(x) * sum((sqrt((x/sum(x))) - sqrt(p))^2)

  expect_equal(a, b)

})

test_that("ks_gof statistic is equal to ks.test statistis (dgof)", {

  a <- as.numeric(ks_gof(x, p, reps = 100)$statistic)

  b <- as.numeric(dgof::ks.test(xx, ecdf(pp))$statistic)

  expect_equal(a, b)

})

test_that("cvm_gof statistic is equal to groupFit statistis (cvmdisc)", {

  a <- as.numeric(cvm_gof(x, p, reps = 100)$statistic)

  b <- cvmdisc::groupFit(breaks, x, distr = "unif")$stats[2, 1]

  expect_equal(a, b)

})


tests/testthat/test-results.R

test_that("chisq_gof equal to chisq.test", {

  x <- c(14, 18, 12, 58)
  p <- c(0.2, 0.2, 0.2, 0.4)

  set.seed(40)
  a <- chisq_gof(x, p)$p.value

  set.seed(40)
  b <- chisq.test(x, p = p, simulate.p.value = TRUE, B = 10000)$p.value

  expect_equal(a, b)

})

test_that("chisq_gof equal to chisq.test", {

  x <- c(134, 128)
  p <- c(0.6, 0.4)

  set.seed(40)
  a <- chisq_gof(x, p)$p.value

  set.seed(40)
  b <- chisq.test(x, p = p, simulate.p.value = TRUE, B = 10000)$p.value

  expect_equal(a, b)

})


test_that("chisq_gof equal to rms_gof for uniform distribution", {

  x <- c(14, 18, 12, 13, 15)
  p <- c(0.2, 0.2, 0.2, 0.2, 0.2)

  set.seed(40)
  a <- chisq_gof(x, p)$p.value

  set.seed(40)
  b <- rms_gof(x, p)$p.value

  expect_equal(a, b)

})

skip_on_cran()

test_that("rms_gof is equal to stored result", {

  x <- c(2114, 2128, 2112)
  p <- c(rep(1/3, 3))


  set.seed(40)
  a <- rms_gof(x, p)$p.value

  b <- 0.966903309669033

  expect_equal(a, b)

})


test_that("ks_gof ~equal to ks.test (from dgof package)", {

  x <- c(14, 18, 15, 28)
  xx <- c(rep(1, 14), rep(2, 18), rep(3, 15), rep(4, 28))
  y <- c(1, 2, 3, 4, 4)
  p <- c(0.2, 0.2, 0.2, 0.4)

  set.seed(284)
  a <- ks_gof(x, p, reps = 2000)$p.value

  set.seed(284)
  b <- dgof::ks.test(xx, ecdf(y), simulate.p.value = TRUE, B = 2000)$p.value

  expect_equal(tolerance = 0.01, a, b)

})

test_that("g_gof is asymtoptically equal to chisq_gof", {

  x <- c(2114, 2128, 2112)
  p <- c(rep(1/3, 3))

  set.seed(300)
  a <- chisq_gof(x, p)$p.value

  set.seed(300)
  b <- g_gof(x, p)$p.value

  expect_equal(a, b)

})

test_that("g_gof is ~equal to G.test (from RVAideMemoire package)", {

  x <- c(214, 228, 212)
  p <- c(rep(1/3, 3))

  set.seed(30)
  a <- g_gof(x, p)$p.value

  b <- 0.707378226223282
  #b <- as.numeric(RVAideMemoire::G.test(x, p)$p.value)

  expect_equal(tolerance = 0.001, a, b)

})

test_that("ft_gof is asymtoptically equal to chisq_gof", {

  x <- c(2114, 2128, 2112)
  p <- c(rep(1/3, 3))

  set.seed(300)
  a <- chisq_gof(x, p)$p.value

  set.seed(300)
  b <- ft_gof(x, p)$p.value

  expect_equal(a, b)

})

test_that("cvm_gof ~equal to groupFit (from cvmdisc package)", {

  x <- c(14, 18, 15, 28)
  p <- c(0.2, 0.2, 0.2, 0.4)
  breaks <- c(0, cumsum(p))

  set.seed(4180)
  a <- cvm_gof(x, p, reps = 10000)$p.value

  set.seed(4180)
  b <- cvmdisc::groupFit(breaks, x, distr = "unif", bootstrap = TRUE, numLoops = 10000)$pvals[2, 1]

  expect_equal(tolerance = 0.01, a, b)

})

test_that("cvm_gof ~equal to cvm.test (from dgof package)", {

  x <- c(1, 4, 11, 4, 0)
  p <- c(0.05, 0.3, 0.3, 0.3, 0.05)


  set.seed(10)
  a <- cvm_gof(x, p, reps = 10000)$p.value

  #xx <- c(rep(1, 1), rep(2, 4), rep(3, 11), rep(4, 4), rep(5, 0))
  #pp <- c(rep(1, 1), rep(2, 6), rep(3, 6), rep(4, 6), rep(5, 1))
  #b <- dgof::cvm.test(xx, ecdf(pp))$p.value

  b <- 0.243733

  expect_equal(tolerance = 0.015, a, b)

})





R/ft_gof.R
#' Simulated Freeman-Tukey (Hellinger-distance) goodness-of-fit test
#'
#' The `ft_gof()` function implements Monte Carlo simulations to calculate p-values
#' based on the Freeman-Tukey statistic for goodness-of-fit tests for discrete
#' distributions. This statistic is also referred to as the Hellinger-distance.
#' Asymptotically, the Freeman-Tukey GOF test is identical to the Chi-squared
#' GOF test, but for smaller n, results may vary significantly.
#'
#' @param x a numeric vector that contains observed counts for each bin/category.
#' @param p a vector of probabilities of the same length of x. An error is given
#'     if any entry of p is negative or if the sum of p does not equal one.
#' @param reps an integer specifying the number of Monte Carlo simulations. The default
#'     is set to 10,000 which may be appropriate for exploratory analysis. A higher
#'     number of simulation should be selected for more precise results.
#'@param tolerance sets an upper bound for rounding errors when evaluating
#'    whether a statistic for a simulation is greater than or equal to the
#'    statistic for the observed data. The default is identical to the tolerance
#'    set for simulations in the `chisq.test` function from the `stats`
#'    package in base R.
#'
#'@return A list with class "htest" containing the following components:
#'
#' \item{statistic}{the value of the Freeman-Tukey test statistic (W2)}
#' \item{p.value}{the simulated p-value for the test}
#' \item{method}{a character string describing the test}
#' \item{data.name}{a character string give the name of the data}
#'
#' @export
#'
#' @examples
#' x <- c(15, 36, 17)
#' p <- c(0.25, 0.5, 0.25)
#'
#'ft_gof(x, p)
#'
#' @importFrom Rcpp sourceCpp
#' @useDynLib discretefit


ft_gof <- function(x, p, reps = 10000, tolerance = 64 * .Machine$double.eps)  {

  errors_x_p(x, p)

  out <- simulate_p(5, x, p, reps, tolerance)

  names(out$statistic) <- "FT"

  val <- list(p.value = out$p_value,
              statistic = out$statistic,
              method = "Simulated Freeman-Tukey goodness-of-fit test",
              data.name = deparse(substitute(x)),
              class = "htest")

  class(val) <- "htest"

  return(val)

}



R/ks_gof.R

#' Simulated Kolmogorov-Smirnov goodness-of-fit test
#'
#' The `ks_gof()` function implements Monte Carlo simulations to calculate p-values
#' based on the Kolmogorov-Smirnov statistic for goodness-of-fit tests for discrete
#' distributions. The p-value expressed by `ks_gof()` is based on a two-sided
#' alternative hypothesis.
#'
#' @param x a numeric vector that contains observed counts for each bin/category.
#' @param p a vector of probabilities of the same length of x. An error is given
#'     if any entry of p is negative or if the sum of p does not equal one.
#' @param reps an integer specifying the number of Monte Carlo simulations. The default
#'     is set to 10,000 which may be appropriate for exploratory analysis. A higher
#'     number of simulation should be selected for more precise results.
#'@param tolerance sets an upper bound for rounding errors when evaluating
#'    whether a statistic for a simulation is greater than or equal to the
#'    statistic for the observed data. The default is identical to the tolerance
#'    set for simulations in the `chisq.test` function from the `stats`
#'    package in base R.
#'
#'@return A list with class "htest" containing the following components:
#'
#' \item{statistic}{the value of the Kolmogorov-Smirnov test statistic}
#' \item{p.value}{the simulated p-value for the test}
#' \item{method}{a character string describing the test}
#' \item{data.name}{a character string give the name of the data}
#'
#' @export
#'
#' @examples
#' x <- c(15, 36, 17)
#' p <- c(0.25, 0.5, 0.25)
#'
#' ks_gof(x, p)
#'
#' @importFrom Rcpp sourceCpp
#' @useDynLib discretefit
#'

ks_gof <- function(x, p, reps = 10000, tolerance = 64 * .Machine$double.eps)  {

  errors_x_p(x, p)

  out <- simulate_p(3, x, p, reps, tolerance)

  names(out$statistic) <- "KS"

  val <- list(p.value = out$p_value,
              statistic = out$statistic,
              method = "Simulated Kolmogorov-Smirnov goodness-of-fit test",
              data.name = deparse(substitute(x)),
              class = "htest")

  class(val) <- "htest"

  return(val)

}


R/cvm_gof.R
#' Simulated Cramer-von Mises goodness-of-fit test
#'
#' The `cvm_gof()` function implements Monte Carlo simulations to calculate p-values
#' based on the Cramer-von Mises statistic (W^2) for goodness-of-fit tests for discrete
#' distributions.
#'
#' @param x a numeric vector that contains observed counts for each bin/category.
#' @param p a vector of probabilities of the same length of x. An error is given
#'     if any entry of p is negative or if the sum of p does not equal one.
#' @param reps an integer specifying the number of Monte Carlo simulations. The default
#'     is set to 10,000 which may be appropriate for exploratory analysis. A higher
#'     number of simulation should be selected for more precise results.
#'@param tolerance sets an upper bound for rounding errors when evaluating
#'    whether a statistic for a simulation is greater than or equal to the
#'    statistic for the observed data. The default is identical to the tolerance
#'    set for simulations in the `chisq.test` function from the `stats`
#'    package in base R.
#'
#' @return A list with class "htest" containing the following components:
#'
#' \item{statistic}{the value of the Cramer-von Mises test statistic (W2)}
#' \item{p.value}{the simulated p-value for the test}
#' \item{method}{a character string describing the test}
#' \item{data.name}{a character string give the name of the data}
#'
#' @export
#'
#' @examples
#' x <- c(15, 36, 17)
#' p <- c(0.25, 0.5, 0.25)
#'
#'cvm_gof(x, p)
#'
#' @importFrom Rcpp sourceCpp
#' @useDynLib discretefit


cvm_gof <- function(x, p, reps = 10000, tolerance = 64 * .Machine$double.eps)  {

  errors_x_p(x, p)

  out <- simulate_p(6, x, p, reps, tolerance)

  names(out$statistic) <- "W2"

  val <- list(p.value = out$p_value,
              statistic = out$statistic,
              method = "Simulated Cramer-von Mises goodness-of-fit test",
              data.name = deparse(substitute(x)),
              class = "htest")

  class(val) <- "htest"

  return(val)

}



R/rms_gof.R

#' Simulated root-mean-square goodness-of-fit test
#'
#' The `rms_gof()` function implements Monte Carlo simulations to calculate p-values
#' based on the root-mean-square statistic for goodness-of-fit tests for discrete
#' distributions.
#'
#' @param x a numeric vector that contains observed counts for each bin/category.
#' @param p a vector of probabilities of the same length of x. An error is given
#'     if any entry of p is negative or if the sum of p does not equal one.
#' @param reps an integer specifying the number of Monte Carlo simulations. The default
#'     is set to 10,000 which may be appropriate for exploratory analysis. A higher
#'     number of simulation should be selected for more precise results.
#'@param tolerance sets an upper bound for rounding errors when evaluating
#'    whether a statistic for a simulation is greater than or equal to the
#'    statistic for the observed data. The default is identical to the tolerance
#'    set for simulations in the `chisq.test` function from the `stats`
#'    package in base R.
#'
#'@return A list with class "htest" containing the following components:
#'
#' \item{statistic}{the value of the root-mean-square test statistic}
#' \item{p.value}{the simulated p-value for the test}
#' \item{method}{a character string describing the test}
#' \item{data.name}{a character string give the name of the data}
#'
#' @export
#'
#' @examples
#' x <- c(15, 36, 17)
#' p <- c(0.25, 0.5, 0.25)
#'
#' rms_gof(x, p)
#'
#' @importFrom Rcpp sourceCpp
#' @useDynLib discretefit
#'
rms_gof <- function(x,
                    p,
                    reps = 10000,
                    tolerance = 64 * .Machine$double.eps)  {

  errors_x_p(x, p)

  out <- simulate_p(1, x, p, reps, tolerance)

  names(out$statistic) <- "RMS"

  val <- list(p.value = out$p_value,
              statistic = out$statistic,
              method = "Simulated root-mean-square goodness-of-fit test",
              data.name = deparse(substitute(x)),
              class = "htest")

  class(val) <- "htest"

  return(val)

}


R/g_gof.R

#' Simulated log-likelihood-ratio (G^2) goodness-of-fit test
#'
#' The `g_gof()` function implements Monte Carlo simulations to calculate p-values
#' based on the log-likelihood-ratio statistic for goodness-of-fit tests for discrete
#' distributions. In this context, the log-likelihood-ratio statistic is often referred
#' to as the G^2 statistic. Asymptotically, the G^2 GOF test is identical to the Chi-squared
#' GOF test, but for smaller n, results may vary significantly.
#'
#' @param x a numeric vector that contains observed counts for each bin/category.
#' @param p a vector of probabilities of the same length of x. An error is given
#'     if any entry of p is negative or if the sum of p does not equal one.
#' @param reps an integer specifying the number of Monte Carlo simulations. The default
#'     is set to 10,000 which may be appropriate for exploratory analysis. A higher
#'     number of simulation should be selected for more precise results.
#'@param tolerance sets an upper bound for rounding errors when evaluating
#'    whether a statistic for a simulation is greater than or equal to the
#'    statistic for the observed data. The default is identical to the tolerance
#'    set for simulations in the `chisq.test` function from the `stats`
#'    package in base R.
#'
#'@return A list with class "htest" containing the following components:
#'
#' \item{statistic}{the value of the log-likelihood-ratio test statistic (G2)}
#' \item{p.value}{the simulated p-value for the test}
#' \item{method}{a character string describing the test}
#' \item{data.name}{a character string give the name of the data}
#'
#' @export
#'
#' @examples
#' x <- c(15, 36, 17)
#' p <- c(0.25, 0.5, 0.25)
#'
#'g_gof(x, p)
#'
#' @importFrom Rcpp sourceCpp
#' @useDynLib discretefit


g_gof <- function(x, p, reps = 10000, tolerance = 64 * .Machine$double.eps)  {

  errors_x_p(x, p)

  out <- simulate_p(4, x, p, reps, tolerance)

  names(out$statistic) <- "G2"

  val <- list(p.value = out$p_value,
              statistic = out$statistic,
              method = "Simulated log-likelihood-ratio goodness-of-fit test",
              data.name = deparse(substitute(x)),
              class = "htest")

  class(val) <- "htest"

  return(val)

}

